hw/ip/otbn/util/otbn-rig

```diff
@@ -54,35 +47,117 @@ def main() -> int:
+                         .format(args.snippets, err))
+        return 1
+
+    program = snippets_to_program(snippets)
+
+    # Dump the assembly output.
+    if args.output is None or args.output == '-':
+        program.dump_asm(sys.stdout)
+    else:
         try:
-            with open(args.asm_output, 'w') as out_file:
+            asm_path = args.output + '.S'
+            with open(asm_path, 'w') as out_file:
                 program.dump_asm(out_file)
         except OSError as err:
-            sys.stderr.write('Failed to open assembly output file {!r}: {}.'
-                             .format(args.asm_output, err))
+            sys.stderr.write('Failed to open asm output file {!r}: {}.'
+                             .format(args.output, err))
             return 1
 
-    # If a linker script was requested, dump that here
-    if args.ld_output is not None:
         try:
-            with open(args.ld_output, 'w') as out_file:
+            ld_path = args.output + '.ld'
+            with open(ld_path, 'w') as out_file:
                 program.dump_linker_script(out_file)
         except OSError as err:
             sys.stderr.write('Failed to open ld script output file {!r}: {}.'
-                             .format(args.ld_output, err))
+                             .format(ld_path, err))
             return 1
 
     return 0
 
 
+def main() -> int:
+    parser = argparse.ArgumentParser()
+    subparsers = parser.add_subparsers()
+
+    gen = subparsers.add_parser('gen', help='Generate a random program')
+    asm = subparsers.add_parser('asm', help='Convert snippets to assembly')
+
+    gen.add_argument('--output', '-o',
+                     default='-',
+                     help=("Path for JSON output of generated snippets. The "
+                           "special path '-' (which is the default) means to "
+                           "write to stdout"))
+    gen.add_argument('--seed', type=int, default=0,
+                     help='Random seed. Defaults to 0.')
+    gen.add_argument('--size', type=int, default=100,
+                     help=('Max number of instructions in stream. '
+                           'Defaults to 100.'))
+    gen.add_argument('--start-addr', type=int, default=0,
+                     help='Reset address. Defaults to 0.')
+    gen.set_defaults(func=gen_main)
+
+    asm.add_argument('--output', '-o',
+                     metavar='out',
+                     help=('Base path for output filenames. Will generate '
+                           'out.S with an assembly listing and out.ld with '
+                           'a linker script. If this is not supplied, the '
+                           'assembly (but no linker script) will be dumped '
+                           'to stdout.'))
+    asm.add_argument('snippets', metavar='path.json',
+                     help=('A JSON file of snippets, as generated by '
+                           'otbn-rig gen.'))
```

Comment:
If you use `argparse.FileType` here, you get support for `-` to read from stdin for free and can do chaining of the `gen` and `asm` commands (`hw/ip/otbn/util/otbn-rig gen --size 10 | hw/ip/otbn/util/otbn-rig asm -o gen -`).

Reply:
Good point! And I've actually made the argument optional as well, so you can just write
```
hw/ip/otbn/util/otbn-rig gen --size 10 | hw/ip/otbn/util/otbn-rig asm
```
if you so choose.